### PR TITLE
merge to master

### PR DESCRIPTION
**Pull request by rMaxiQp**
add **Role** column in **CourseOffering** table
add **Role** feature to ACL, default role set as "Student"

### CODE REVIEW

File: db/db.js [hunk 1]
```diff
@@ -60,14 +55,13 @@ async function getMediaIdsByCourseOfferingId(courseOfferingId) {
 
 function getPlaylistByCourseOfferingId(courseOfferingId) {
   return sequelize.query(
-   'SELECT mst.videoLocalLocation, mst.srtFileLocation, M.siteSpecificJSON, mst.mediaId \
+   'SELECT mst.videoLocalLocation, mst.srtFileLocation, M.siteSpecificJSON \
```
Comment (chirantan31) [inline db/db.js]: add the mst.mediaId

File: db/db.js [hunk 2]
```diff
@@ -60,14 +55,13 @@ async function getMediaIdsByCourseOfferingId(courseOfferingId) {
...
     FROM MSTranscriptionTasks AS mst \
     INNER JOIN TaskMedia as tm on tm.taskId = mst.id \
     INNER JOIN Media as M on tm.mediaId = M.id \
     INNER JOIN CourseOfferingMedia as com on com.mediaId = M.id \
-    WHERE com.courseOfferingId = ? \
-    ORDER BY M.id',
-   { replacements: [ courseOfferingId ], type: sequelize.QueryTypes.SELECT}).catch(err => perror(err)); /* raw query */
+    WHERE com.courseOfferingId = ?',
```
Comment (chirantan31) [inline db/db.js]: add the ORDER BY clause

File: db/db.js [hunk 3]
```diff
@@ -259,7 +253,7 @@ function getUserByGoogleId(profileId) {
         } else {
             return null;
         }
-    }).catch(err => perror(err));
+    }).catch(err => {throw new Error(err.message)});
```
Comment (chirantan31) [inline db/db.js]: Who's catching these errors that are thrown?

File: router/routes/captions.js [hunk 4]
```diff
@@ -1,24 +1,28 @@
...
-  response.writeHead(200, {
-    'Content-Type': 'application/json'
-  });
-
-  var index = parseInt(request.params.index);
-  response.end(JSON.stringify({ captions: captions[index] }));
-});
-
-module.exports = router;
+/** Copyright 2015 Board of Trustees of University of Illinois
```
Comment (chirantan31) [inline router/routes/captions.js]: Do we use this file anymore?
If not delete it

File: router/routes/changePassword.js [hunk 5]
```diff
@@ -26,25 +26,27 @@ router.get('/changePassword', function (request, response) {
     renderWithPartial(Mustache.getMustacheTemplate('changePassword.mustache'), request, response);
   } else {
```
Comment (chirantan31) [inline router/routes/changePassword.js]: I don't think we use the changePassword.js flow anymore, lets get rid of it

File: router/routes/download.js [hunk 6]
```diff
@@ -21,6 +21,7 @@ router.post('/download', function (request, response) {
 });
```
Comment (chirantan31) [inline router/routes/download.js]: Is this route used anymore?
If not, delete

File: router/routes/first.js [hunk 7]
```diff
@@ -17,6 +17,7 @@ var client_api = require('../../db/db');
```
Comment (chirantan31) [inline router/routes/first.js]: Delete if not in use

File: router/routes/signup.js [hunk 8]
```diff
@@ -152,7 +152,7 @@ router.post('/signup/submit', function (request, response) {
```
Comment (chirantan31) [inline router/routes/signup.js]: If this signup flow isn't used, delete

File: router/routes/search.js [hunk 9]
```diff
@@ -16,6 +16,7 @@ const db = require('../../db/db');
 const invalidClassHTML = "<p>Could not find the requested page.<\p> <a href=\"/\">Click here to return to the home page.</a>";
```
Comment (chirantan31) [inline router/routes/search.js]: Dont' think this flow is used anymore, confirm and delete

File: router/routes/scraper.js [hunk 10]
```diff
@@ -11,6 +11,7 @@ var scraper = require('../../modules/scraper_utils');
```
Comment (chirantan31) [inline router/routes/scraper.js]: This is kind of an admin page, all the routes are only explicitly called by me, so you can remove the revert this file

File: router/routes/resetPassword.js [hunk 11]
```diff
@@ -48,17 +48,18 @@ router.get('/resetPassword', function (request, response) {
 router.post('/resetPassword/submit', function (request, response) {
```
Comment (chirantan31) [inline router/routes/resetPassword.js]: I think this flow isn't used anymore, if not delete

File: router/routes/queue.js [hunk 12]
```diff
@@ -13,6 +13,7 @@ const db = require('../../db/db');
```
Comment (chirantan31) [inline router/routes/queue.js]: I think this flow isn't used anymore, if not delete

File: router/routes/progress.js [hunk 13]
```diff
@@ -10,6 +10,7 @@ const client_api = require('./db');
 
 var progressMustache = fs.readFileSync(mustachePath + 'progress.mustache').toString();
```
Comment (chirantan31) [inline router/routes/progress.js]: I think this flow isn't used anymore, if not delete

File: router/routes/permission.js [hunk 14]
```diff
@@ -3,19 +3,26 @@
  *     https://www.npmjs.com/package/acl
  */
```
Comment (chirantan31) [inline router/routes/permission.js]: The scope of permissions have to be extended more,
Depending on the role column you've added to the CourseOfferin table,
Based on that a user might have to be given individual access to a certain courseoffering etc.

Lets discuss this in detail

File: db/db.js (hunk 3 above)
Comment (rMaxiQp) [reply db/db.js]: throw error when accessing db, let caller catch errors. By doing so, we may have an easier time for logging (no need of passing userInfo around all the time)

File: router/routes/permission.js (hunk 14 above)
Comment (rMaxiQp) [reply router/routes/permission.js]: To check if user has a permission to a certain course, let's handle it through ACL. To filter out course with certain permission level requirements, use the attribute stored in the database.

**Comment by rMaxiQp** [comment]
Close hanging PR.

**Pull request closed without merging**